File: jscomp/others/belt_MutableSetInt.ml
# 1 "others/setm.cppo.ml"
(* Copyright (C) 2017 Hongbo Zhang, Authors of ReScript
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

(** This module is [`Belt.MutableSet`]() specialized with key type to be a primitive type.
    It is more efficient in general, the  API is the same with [`Belt_MutableSet`]() except its key type is fixed,
    and identity is not needed(using the built-in one)
*)

# 31 "others/setm.cppo.ml"
module I = Belt_internalSetInt
module S = Belt_SortArrayInt
# 39 "others/setm.cppo.ml"
module N = Belt_internalAVLset
module A = Belt_Array

type value = I.value
(** The type of the set elements. *)


type t = {
  mutable data : I.t
}
(** The type of sets. *)


let rec remove0 nt (x : value)=
  let k = nt.N.value in
  if x = k then
    let {N.left = l; right = r} = nt in
    match l, r with
    | None, _ -> r
    | _, None -> l
    | Some _,  Some nr ->
      nt.right <- (N.removeMinAuxWithRootMutate nt nr);
      Some (N.balMutate nt)
  else
    begin
      if x < k then
        match nt.left with
        | None -> Some nt
        | Some l ->
          nt.left <- (remove0 l x );
          Some (N.balMutate nt)
      else
        match nt.right with
        | None -> Some nt
        | Some r ->
          nt.right <- (remove0 r x);
          Some (N.balMutate nt)
    end

let remove d v =
  let oldRoot = d.data in
  match oldRoot with
  | None -> ()
  | Some oldRoot2 ->
  let newRoot = remove0 oldRoot2 v in
  if newRoot != oldRoot then
    d.data <- newRoot

let rec removeMany0 t xs i len  =
  if i < len then
    let ele = A.getUnsafe xs i in
    let u = remove0 t ele in
    match u with
    | None -> None
    | Some t -> removeMany0 t xs (i+1) len
  else Some t


let removeMany  (d : t) xs =
  let oldRoot = d.data in
  match oldRoot with
  | None -> ()
  | Some nt ->
    let len = A.length xs in
    d.data <- removeMany0 nt xs 0 len

let rec removeCheck0  nt (x : value) removed =
  let k = nt.N.value in
  if x = k then
    let () = removed .contents<- true in
    let {N.left = l; right = r} = nt in
    match l, r with
    | None, _ -> r
    | _ , None -> l
    | Some _,  Some nr ->
      nt.right <- (N.removeMinAuxWithRootMutate nt nr);
      Some (N.balMutate nt)
  else
    begin
      if x < k then
        match nt.left with
        | None -> Some nt
        | Some l ->
          nt.left <- (removeCheck0  l x removed);
          Some (N.balMutate nt)
      else
        match nt.right with
        | None -> Some nt
        | Some r ->
          nt.right <- (removeCheck0  r x removed);
          Some (N.balMutate nt)
    end



let removeCheck  (d :  t) v =
  let oldRoot = d.data in
  match oldRoot with
  | None -> false
  | Some oldRoot2 ->
    let removed = ref false in
    let newRoot = removeCheck0  oldRoot2 v removed in
    if newRoot != oldRoot then
      d.data <- newRoot ;
    removed.contents


let rec addCheck0  t (x : value) added  =
  match t with
  | None ->
    added .contents<- true;
    N.singleton x
  | Some nt ->
    let k = nt.N.value in
    if x = k then t
    else
      let {N.left = l; right =  r} = nt in
      (if x < k then
         let ll = addCheck0  l x added in
         nt.left <- ll
       else
         nt.right <- (addCheck0 r x added );
      );
      Some (N.balMutate nt)

let addCheck (m :  t) e =
  let oldRoot = m.data in
  let added = ref false in
  let newRoot = addCheck0 oldRoot e added in
  if newRoot != oldRoot then
    m.data <- newRoot;
  added.contents

let add d k =
  let oldRoot = d.data in
  let v = I.addMutate oldRoot k in
  if v != oldRoot then
     d.data <- v


let addArrayMutate t  xs =
  let v = ref t in
  for i = 0 to A.length xs - 1 do
    v.contents<- I.addMutate v.contents (A.getUnsafe xs i)
  done ;
  v.contents

let mergeMany d arr =
  d.data <- addArrayMutate (d.data) arr



let make  () = {data = None}

let isEmpty d =
  N.isEmpty (d.data)

let minimum d =
  N.minimum (d.data)

let minUndefined d =
  N.minUndefined (d.data)

let maximum d = N.maximum (d.data)

let maxUndefined d = N.maxUndefined (d.data)

let forEachU d f = N.forEachU (d.data) f
let forEach d f = forEachU d (fun[@bs] a -> f a)

let reduceU d acc cb = N.reduceU (d.data) acc cb
let reduce d acc cb = reduceU d acc (fun[@bs] a b -> cb a b)

let everyU d p = N.everyU (d.data) p
let every d p = everyU d (fun[@bs] a -> p a)
let someU d p = N.someU (d.data) p
let some d p = someU d (fun [@bs] a -> p a)
let size d =
  N.size (d.data)
let toList d =
  N.toList (d.data)
let toArray d =
  N.toArray (d.data)


let fromSortedArrayUnsafe xs =
  {data = N.fromSortedArrayUnsafe xs}

let checkInvariantInternal d =
  N.checkInvariantInternal (d.data)



let fromArray xs =
  { data = I.fromArray xs}

let cmp d0 d1 =
  I.cmp (d0.data) (d1.data)
let eq d0 d1 =
  I.eq (d0.data) (d1.data)
let get d x =
  I.get (d.data) x
let getUndefined d x =
  I.getUndefined (d.data) x
let getExn d x =
  I.getExn (d.data) x

let split d  key =
  let arr = N.toArray (d.data) in
  let i = S.binarySearch arr key   in
  let len = A.length arr in
  if i < 0 then
    let next = - i -1 in
    (
      {data = N.fromSortedArrayAux arr 0 next}
    ,

      {data = N.fromSortedArrayAux arr next (len - next)}
    ), false
  else
    (
      {data = N.fromSortedArrayAux arr 0 i}
    ,

      {data = (N.fromSortedArrayAux arr (i+1) (len - i - 1))}
      ), true

let keepU d p =
  {data = (N.keepCopyU (d.data) p )}
let keep d p = keepU d (fun[@bs] a -> p a)

let partitionU d p =
  let a , b = N.partitionCopyU (d.data) p in
  {data = a}, {data = b}
let partition d p = partitionU d (fun[@bs] a -> p a)

let subset a b = I.subset  a.data b.data
let intersect dataa datab  =
  let dataa, datab = dataa.data, datab.data in
    match dataa, datab with
    | None, _ -> make ()
    | _, None -> make ()
    | Some dataa0, Some datab0 ->
    let sizea, sizeb =
        N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore @@ N.fillArray dataa0 0 tmp ;
    ignore @@ N.fillArray datab0 sizea tmp;
    if ((A.getUnsafe tmp (sizea - 1) <
        A.getUnsafe tmp sizea))
      ||
      (
      (A.getUnsafe tmp (totalSize - 1) <
      A.getUnsafe tmp 0)
      )
       then make ()
    else
    let tmp2 = A.makeUninitializedUnsafe (Pervasives.min sizea sizeb) in
    let k = S.intersect tmp 0 sizea tmp sizea sizeb tmp2 0  in
    {data = (N.fromSortedArrayAux tmp2 0 k)}

let diff dataa datab : t =
  let dataa, datab = dataa.data, datab.data in
  match dataa, datab with
  | None, _ -> make ()
  | _, None -> {data = N.copy dataa}
  | Some dataa0, Some datab0 ->
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore @@ N.fillArray dataa0 0 tmp ;
    ignore @@ N.fillArray datab0 sizea tmp;
    if ( (A.getUnsafe tmp (sizea - 1)) <
        (A.getUnsafe tmp sizea))
      ||
      (A.getUnsafe tmp (totalSize - 1)
      < A.getUnsafe tmp 0)
       then {data = N.copy dataa}
    else
    let tmp2 = A.makeUninitializedUnsafe sizea in
    let k = S.diff tmp 0 sizea tmp sizea sizeb tmp2 0  in
    {data = (N.fromSortedArrayAux tmp2 0 k)}

let union (dataa : t)  (datab : t) : t =
  let dataa, datab = dataa.data, datab.data in
   match dataa, datab with
  | None, _ -> {data = (N.copy datab)}
  | _, None -> {data = (N.copy dataa)}
  | Some dataa0, Some datab0
    ->
    let sizea, sizeb = N.lengthNode dataa0, N.lengthNode datab0 in
    let totalSize = sizea + sizeb in
    let tmp = A.makeUninitializedUnsafe totalSize in
    ignore @@ N.fillArray dataa0 0 tmp ;
    ignore @@ N.fillArray datab0 sizea tmp ;
    if
      (A.getUnsafe tmp (sizea - 1) <
      A.getUnsafe tmp sizea)  then
      {data = (N.fromSortedArrayAux tmp 0 totalSize) }
    else
      let tmp2 = A.makeUninitializedUnsafe totalSize in
      let k = S.union tmp 0 sizea tmp sizea sizeb tmp2 0  in
      {data = (N.fromSortedArrayAux tmp2 0 k) }

let has d x = I.has (d.data) x

let copy d = {data = (N.copy (d.data))}

File: lib/es6/Belt_MutableSetInt.js


import * as Primitive_int from "./Primitive_int.js";
import * as Belt_SortArrayInt from "./Belt_SortArrayInt.js";
import * as Belt_internalAVLset from "./Belt_internalAVLset.js";
import * as Belt_internalSetInt from "./Belt_internalSetInt.js";

function remove0(nt, x) {
  let k = nt.v;
  if (x === k) {
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (x < k) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = remove0(l$1, x);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = remove0(r$1, x);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, v) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let newRoot = remove0(oldRoot, v);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function removeMany0(_t, xs, _i, len) {
  while (true) {
    let i = _i;
    let t = _t;
    if (i >= len) {
      return t;
    }
    let ele = xs[i];
    let u = remove0(t, ele);
    if (u === undefined) {
      return;
    }
    _i = i + 1 | 0;
    _t = u;
    continue;
  };
}

function removeMany(d, xs) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  let len = xs.length;
  d.data = removeMany0(oldRoot, xs, 0, len);
}

function removeCheck0(nt, x, removed) {
  let k = nt.v;
  if (x === k) {
    removed.contents = true;
    let l = nt.l;
    let r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt_internalAVLset.removeMinAuxWithRootMutate(nt, r);
        return Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (x < k) {
    let l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeCheck0(l$1, x, removed);
      return Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  let r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeCheck0(r$1, x, removed);
    return Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function removeCheck(d, v) {
  let oldRoot = d.data;
  if (oldRoot === undefined) {
    return false;
  }
  let removed = {
    contents: false
  };
  let newRoot = removeCheck0(oldRoot, v, removed);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
  }
  return removed.contents;
}

function addCheck0(t, x, added) {
  if (t !== undefined) {
    let k = t.v;
    if (x === k) {
      return t;
    }
    let l = t.l;
    let r = t.r;
    if (x < k) {
      let ll = addCheck0(l, x, added);
      t.l = ll;
    } else {
      t.r = addCheck0(r, x, added);
    }
    return Belt_internalAVLset.balMutate(t);
  }
  added.contents = true;
  return Belt_internalAVLset.singleton(x);
}

function addCheck(m, e) {
  let oldRoot = m.data;
  let added = {
    contents: false
  };
  let newRoot = addCheck0(oldRoot, e, added);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
  }
  return added.contents;
}

function add(d, k) {
  let oldRoot = d.data;
  let v = Belt_internalSetInt.addMutate(oldRoot, k);
  if (v !== oldRoot) {
    d.data = v;
    return;
  }
  
}

function addArrayMutate(t, xs) {
  let v = t;
  for (let i = 0, i_finish = xs.length; i < i_finish; ++i) {
    v = Belt_internalSetInt.addMutate(v, xs[i]);
  }
  return v;
}

function mergeMany(d, arr) {
  d.data = addArrayMutate(d.data, arr);
}

function make() {
  return {
    data: undefined
  };
}

function isEmpty(d) {
  let n = d.data;
  return n === undefined;
}

function minimum(d) {
  return Belt_internalAVLset.minimum(d.data);
}

function minUndefined(d) {
  return Belt_internalAVLset.minUndefined(d.data);
}

function maximum(d) {
  return Belt_internalAVLset.maximum(d.data);
}

function maxUndefined(d) {
  return Belt_internalAVLset.maxUndefined(d.data);
}

function forEach(d, f) {
  Belt_internalAVLset.forEach(d.data, f);
}

function reduce(d, acc, cb) {
  return Belt_internalAVLset.reduce(d.data, acc, cb);
}

function every(d, p) {
  return Belt_internalAVLset.every(d.data, p);
}

function some(d, p) {
  return Belt_internalAVLset.some(d.data, p);
}

function size(d) {
  return Belt_internalAVLset.size(d.data);
}

function toList(d) {
  return Belt_internalAVLset.toList(d.data);
}

function toArray(d) {
  return Belt_internalAVLset.toArray(d.data);
}

function fromSortedArrayUnsafe(xs) {
  return {
    data: Belt_internalAVLset.fromSortedArrayUnsafe(xs)
  };
}

function checkInvariantInternal(d) {
  Belt_internalAVLset.checkInvariantInternal(d.data);
}

function fromArray(xs) {
  return {
    data: Belt_internalSetInt.fromArray(xs)
  };
}

function cmp(d0, d1) {
  return Belt_internalSetInt.cmp(d0.data, d1.data);
}

function eq(d0, d1) {
  return Belt_internalSetInt.eq(d0.data, d1.data);
}

function get(d, x) {
  return Belt_internalSetInt.get(d.data, x);
}

function getUndefined(d, x) {
  return Belt_internalSetInt.getUndefined(d.data, x);
}

function getExn(d, x) {
  return Belt_internalSetInt.getExn(d.data, x);
}

function split(d, key) {
  let arr = Belt_internalAVLset.toArray(d.data);
  let i = Belt_SortArrayInt.binarySearch(arr, key);
  let len = arr.length;
  if (i >= 0) {
    return [
      [
        {
          data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, i)
        },
        {
          data: Belt_internalAVLset.fromSortedArrayAux(arr, i + 1 | 0, (len - i | 0) - 1 | 0)
        }
      ],
      true
    ];
  }
  let next = (-i | 0) - 1 | 0;
  return [
    [
      {
        data: Belt_internalAVLset.fromSortedArrayAux(arr, 0, next)
      },
      {
        data: Belt_internalAVLset.fromSortedArrayAux(arr, next, len - next | 0)
      }
    ],
    false
  ];
}

function keep(d, p) {
  return {
    data: Belt_internalAVLset.keepCopy(d.data, p)
  };
}

function partition(d, p) {
  let match = Belt_internalAVLset.partitionCopy(d.data, p);
  return [
    {
      data: match[0]
    },
    {
      data: match[1]
    }
  ];
}

function subset(a, b) {
  return Belt_internalSetInt.subset(a.data, b.data);
}

function intersect(dataa, datab) {
  let dataa$1 = dataa.data;
  let datab$1 = datab.data;
  if (dataa$1 === undefined) {
    return {
      data: undefined
    };
  }
  if (datab$1 === undefined) {
    return {
      data: undefined
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(dataa$1);
  let sizeb = Belt_internalAVLset.lengthNode(datab$1);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa$1, 0, tmp);
  Belt_internalAVLset.fillArray(datab$1, sizea, tmp);
  if (tmp[sizea - 1 | 0] < tmp[sizea] || tmp[totalSize - 1 | 0] < tmp[0]) {
    return {
      data: undefined
    };
  }
  let tmp2 = new Array(Primitive_int.min(sizea, sizeb));
  let k = Belt_SortArrayInt.intersect(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0);
  return {
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function diff(dataa, datab) {
  let dataa$1 = dataa.data;
  let datab$1 = datab.data;
  if (dataa$1 === undefined) {
    return {
      data: undefined
    };
  }
  if (datab$1 === undefined) {
    return {
      data: Belt_internalAVLset.copy(dataa$1)
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(dataa$1);
  let sizeb = Belt_internalAVLset.lengthNode(datab$1);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa$1, 0, tmp);
  Belt_internalAVLset.fillArray(datab$1, sizea, tmp);
  if (tmp[sizea - 1 | 0] < tmp[sizea] || tmp[totalSize - 1 | 0] < tmp[0]) {
    return {
      data: Belt_internalAVLset.copy(dataa$1)
    };
  }
  let tmp2 = new Array(sizea);
  let k = Belt_SortArrayInt.diff(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0);
  return {
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function union(dataa, datab) {
  let dataa$1 = dataa.data;
  let datab$1 = datab.data;
  if (dataa$1 === undefined) {
    return {
      data: Belt_internalAVLset.copy(datab$1)
    };
  }
  if (datab$1 === undefined) {
    return {
      data: Belt_internalAVLset.copy(dataa$1)
    };
  }
  let sizea = Belt_internalAVLset.lengthNode(dataa$1);
  let sizeb = Belt_internalAVLset.lengthNode(datab$1);
  let totalSize = sizea + sizeb | 0;
  let tmp = new Array(totalSize);
  Belt_internalAVLset.fillArray(dataa$1, 0, tmp);
  Belt_internalAVLset.fillArray(datab$1, sizea, tmp);
  if (tmp[sizea - 1 | 0] < tmp[sizea]) {
    return {
      data: Belt_internalAVLset.fromSortedArrayAux(tmp, 0, totalSize)
    };
  }
  let tmp2 = new Array(totalSize);
  let k = Belt_SortArrayInt.union(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0);
  return {
    data: Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function has(d, x) {
  return Belt_internalSetInt.has(d.data, x);
}

function copy(d) {
  return {
    data: Belt_internalAVLset.copy(d.data)
  };
}

let forEachU = forEach;

let reduceU = reduce;

let everyU = every;

let someU = some;

let keepU = keep;

let partitionU = partition;

export {
  make,
  fromArray,
  fromSortedArrayUnsafe,
  copy,
  isEmpty,
  has,
  add,
  addCheck,
  mergeMany,
  remove,
  removeCheck,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toList,
  toArray,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
}
/* No side effect */
